MQClient/backends/pulsar.py

```diff
@@ -0,0 +1,189 @@
+    """
+
+    def __init__(self, address: str, topic: str) -> None:
+        super().__init__(address, topic)
+        self.producer = None  # type: pulsar.Producer
+
+    def connect(self):
+        """Connect to producer."""
+        super().connect()
+        self.producer = self.client.create_producer(self.topic)
+
+
+class PulsarSub(Pulsar):
+    """Wrapper around pulsar.Consumer.
+
+    Extends:
+        Pulsar
+    """
+
+    def __init__(self, address: str, topic: str) -> None:
+        super().__init__(address, topic)
+        self.consumer = None  # type: pulsar.Consumer
+        self.subscription_name = f'{self.topic}-subscription'  # single shared subscription
+        self.prefetch = 1
+
+    def connect(self):
+        """Connect to subscriber."""
+        super().connect()
+        self.consumer = self.client.subscribe(self.topic,
+                                              self.subscription_name,
+                                              receiver_queue_size=self.prefetch,
+                                              consumer_type=pulsar.ConsumerType.Shared,
+                                              initial_position=pulsar.InitialPosition.Earliest)
+
+    def close(self):
+        """Close client and redeliver any unacknowledged messages."""
+        if self.consumer:
+            self.consumer.redeliver_unacknowledged_messages()
+        super().close()
+
+# Interface Methods
+
+
+def create_pub_queue(address: str, name: str) -> PulsarPub:
+    """Create a publishing queue."""
+    q = PulsarPub(address, name)
+    q.connect()
+    return q
+
+
+def create_sub_queue(address: str, name: str, prefetch: int = 1) -> PulsarSub:
+    """Create a subscription queue."""
+    q = PulsarSub(address, name)
+    q.prefetch = prefetch
+    q.connect()
+    return q
+
+
+def send_message(queue: PulsarPub, msg: bytes) -> None:
+    """Send a message on a queue."""
+    if not queue.producer:
+        raise RuntimeError("queue is not connected")
+
+    queue.producer.send(msg)
+
+
+def get_message(queue: PulsarSub, timeout_millis: int = 100) -> typing.Optional[Message]:
+    """Get a single message from a queue.
+
+    To endlessly block until a message is available, set `timeout_millis=None`.
+    """
+    if not queue.consumer:
+        raise RuntimeError("queue is not connected")
+
+    while True:
+        try:
+            msg = queue.consumer.receive(timeout_millis=timeout_millis)
+            if msg:
+                message_id, data = msg.message_id(), msg.data()
+                if message_id and data:
+                    return Message(message_id, data)
+            return None
+
+        except Exception as e:
+            if str(e) == "Pulsar error: TimeOut":  # pulsar isn't a fan of derived Exceptions
+                return None
+            if str(e) == "Pulsar error: AlreadyClosed":
+                queue.connect()
+                continue
+            raise
+
+        break  # poor man's do-while loop
```

Comment:
Is the `break` at the end reachable?
It seems like the try block would return a Message object, return None, or exception out.
It seems like the except block would return None, jump back up to while True, or exception out by raising the original exception.

Reply:
Good catch. I'm surprised one off my linters wasn't able to catch this.